TensorLib/TensorData.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: John Tristan, Paul Govereau, Sean McLaughlin
+-/
+import Mathlib.Tactic
+import TensorLib.NonEmptyList
+import TensorLib.TensorElement
+
+namespace TensorLib
+
+/-!
+This file defines TensorData, the raw data underlying a Tensor.
+We are representing it today as nested linked lists. We predict
+this will be easier to reason about, but will be slow to evaluate.
+We intend to have a more compact and faster-to-access representation
+in addition to this one in the future for execution.
+-/
+
+-- I tried making this implicit with {} but I needed to add @ everwhere
+-- and give it the argument anyway. Maybe there is a way to infer at least
+-- within a section
+variable (a: Type)
+
+inductive TensorData where
+| td_root (xs: List a)
+| td_node (xs: NonEmptyList TensorData)
+
+instance : Inhabited (TensorData a) where
+  default := .td_root []
+
+def td_node! (xs: List (TensorData a)): TensorData a :=
+  .td_node (NonEmptyList.fromList! xs)
+
+-- theorem tensorDataInduction (P : TensorData a -> Prop)
+--  (hroot : ∀ (xs: List a), P (.td_root xs))
+--  (hnode : ∀ (xs: List (TensorData _)), (∀ x, x ∈ xs -> P x) -> P (.td_node xs))
+--  (x: TensorData a): P x :=
+--  match x with
+--   | .td_root xs => hroot xs
+--   | .td_node xs => hnode xs (fun x _ => tensorDataInduction P hroot hnode x)
+
+abbrev Shape := List Nat
+
+-- TODO: Make tail-recursive, perhaps when Lean will eliminite tail calls in mutual-recursion
+def TensorData.hasShape (x: TensorData a) (shape: Shape): Bool :=
+  match x, shape with
+  | .td_root y, List.cons n .nil => List.length y == n
+  | .td_node ys, List.cons n shape' =>
+      ys.length == n &&
+      -- `attach` is used in proofs below, e.g. HashShapeOk
+      List.all ys.toList.attach (fun ⟨ y, _ ⟩ => hasShape y shape')
+  | _, _ => false
+
+lemma hasShapeNil (x: TensorData a): ! (x.hasShape a .nil) := by
+  unfold TensorData.hasShape
+  simp
+
+lemma hasShapeRoot (xs: List a)
+  (H: (TensorData.td_root xs).hasShape a (.cons c cs)) :
+  cs = .nil ∧ List.length xs = c := by
+  revert H
```

Comment:
strange revert. You probably need to learn to manipulate formulas in the context directly

Reply:
Tried a few variations with just messing with H, and none were so succinct. Similar to the other cases below. e.g. when I revert A \/ B, and simp, I can intros A and B separately.